include/swift/Basic/PriorityQueue.h

```diff
@@ -0,0 +1,161 @@
+///
+/// The `Node` type parameter represents a reference to a list node.
+/// Conceptually, a `Node` value is either null or a reference to an
+/// object with an abstract sort value and a `next` reference
+/// (another `Node` value).
+///
+/// A null reference can be created by explicitly default-constructing
+/// the `Node` type, e.g. with `Node()`.  Converting a `Node` value
+/// contextually to `bool` tests whether the node is a null reference.
+/// `Node` values can be compared with the `==` and `!=` operators,
+/// and equality with `Node()` is equivalent to a `bool` conversion.
+/// These conditions are designed to allow pointer types to be used
+/// directly, but they also permit other types.  `ListMerger` is not
+/// currently written to support smart pointer types efficiently,
+/// however.
+///
+/// The sort value and `next` reference are not accessed directly;
+/// instead, they are accessed with `static` functions on the
+/// `NodeTraits` type parameter:
+///
+/// ```
+///   /// Return the current value of the next reference.
+///   static Node getNext(Node n);
+///
+///   /// Set the current value of the next reference.
+///   static void setNext(Node n, Node next);
+///
+///   /// Total number of priority buckets.
+///   enum { prioritiesCount = ... };
+///
+///   /// Returns priority of the Node as value between 0 and `prioritiesCount-1`.
+///   /// Smaller indices have higher priority.
+///   static int getPriorityIndex(Node);
+/// ```
+///
+/// All nodes are stored in a single linked list, sorted by priority.
+/// Within the same priority jobs are sorted in the FIFO order.
+///
+template <class Node, class NodeTraits>
+class PriorityQueue {
+private:
+  /// Head of the linked list.
+  Node head;
+  /// Last node of the corresponding priority, or null if no nodes of that
+  /// priority exist.
+  Node tails[NodeTraits::prioritiesCount];
+
+public:
+  PriorityQueue() : head(), tails{} {}
+
+  /// Add a single node to this queue.
+  ///
+  /// The next reference of the node will be overwritten and does not
+  /// need to be meaningful.
+  ///
+  /// The relative order of the existing nodes in the queue will not change,
+  /// and if there are nodes in the current list which compare equal
+  /// to the new node, it will be inserted after them.
+  void enqueue(Node newNode) {
+    assert(newNode && "inserting a null node");
+    int priorityIndex = NodeTraits::getPriorityIndex(newNode);
+    enqueueRun(priorityIndex, newNode, newNode);
+  }
+
+  /// Add a chain of nodes of mixed priorities to this queue.
+  void enqueueContentsOf(Node otherHead) {
+    Node runHead = otherHead;
+    while (runHead) {
+      int priorityIndex = NodeTraits::getPriorityIndex(runHead);
+
+      // Find run of jobs of the same priority
+      Node runTail = runHead;
+      Node next = NodeTraits::getNext(runTail);
+      while (true) {
+        if (!next) break;
+        if (NodeTraits::getPriorityIndex(next) != priorityIndex) break;
```

Comment:
It's pretty minor, but you could save this as `nextRunPriorityIndex` and then overwrite `priorityIndex` when you start the next run.  Getting the priority isn't a totally trivial operation, so it might be worthwhile.